internal/updatecheck/update.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package updatecheck
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	"github.com/fatih/color"
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+
+	"sigs.k8s.io/krew/internal/version"
+)
+
+const (
+	githubVersionURL = "https://api.github.com/repos/kubernetes-sigs/krew/releases/latest"
+
+	upgradeNotification = `A newer version of krew is available (%s -> %s). 
+Run "kubectl krew upgrade" to get the newest version!`
+)
+
+// for testing
+var versionURL = githubVersionURL
+
+// CheckVersion returns a notification message to inform the user
+// about a new version of krew, or an empty string. The notification
+// will only be emitted once a day.
+func CheckVersion(basePath string) (string, error) {
+	f := filepath.Join(basePath, "last_update_check")
+	if lastCheck := loadTimestamp(f); time.Since(lastCheck).Hours() <= 24 {
+		klog.V(3).Info("Last check was recently, skipping update check")
+		return "", nil
+	}
+
+	latestTag, err := fetchLatestTag()
+	if err != nil {
+		return "", errors.Wrapf(err, "could not determine latest tag")
+	}
+	saveTimestamp(f)
+
+	if version.GitTag() == latestTag {
```

Comment:
also worth keeping in mind that sometimes version.GitTag() might return malformed or empty string that isn't a semver.

if we actually parse that higher in the stack, we don't even have to `fetchLatestTag()`.

Reply:
Do you mean to say that the update check should not run for development builds? That would make perfect sense.

Reply:
Why not? If I want to be in the edge, I would care updating even more than a regular user.

Reply:
Why not:
- if you are developing krew, you are usually ahead of the latest release. So the "upgrade" notification would recommend a downgrade :)
- the CI runs non-release builds of krew in integration tests. We don't need upgrade notifications there.

Reply:
> Do you mean to say that the update check should not run for development builds? That would make perfect sense.

Not just development builds.
For example, until recently homebrew formula did not have `-ldflags` that stamp the GitTag into the binary.
Also think of people running `go get sigs.k8s.io/krew/cmd/krew`, it will not have a GitTag in it.

For those cases, just fail to parse version as semver, and move on.

> * if you are developing krew, you are usually ahead of the latest release. So the "upgrade" notification would recommend a downgrade :)

thats why (1) parse strings into SemVer (2) use semver.LessThan to compare instead of  `!=`.